### PR TITLE
[wip] Always log train metrics on last batch

### PR DESCRIPTION
**Pull request by ozen**
## What does this PR do?

Currently, training metrics are logged on batch end if one of the three conditions is satisfied:
* row_log_interval is reached
* trainer.should_stop is True
* fast_dev_run is True

Sometimes the default value for row_log_interval (50) is less then the number of batches. In that case, training metrics are never logged. This can't be desired. To force the best practice, this PR adds two more situations when the training metrics are logged:
* is_last_batch is True
* batch_idx + 1 == num_training_batches

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -474,7 +474,13 @@ def run_training_epoch(self):
             # -----------------------------------------
             # SAVE METRICS TO LOGGERS
             # -----------------------------------------
-            self.trainer.logger_connector.save_train_loop_metrics_to_loggers(batch_idx, batch_output)
+            should_log_metrics = ((batch_idx + 1) % self.trainer.row_log_interval == 0
```
Comment (SkafteNicki) [inline pytorch_lightning/trainer/training_loop.py:477]: The logic should be moved back to the `save_train_loop_metrics_to_loggers` function to keep the training loop as clean as possible.

Comment (ozen) [reply pytorch_lightning/trainer/training_loop.py]: Done.

**Comment by SkafteNicki** [comment]
@ozen can you add a test which currently would fail to log something, but will log with the changes you made?

**Comment by awaelchli** [comment]
If the number of batches in one epoch is much smaller than the row_log_interval, it can be a bottleneck with logging and there is no way to prevent it, as increasing row_log_interval has no effect in this case.

**Comment by ozen** [comment]
> 
> 
> If the number of batches in one epoch is much smaller than the row_log_interval, it can be a bottleneck with logging and there is no way to prevent it, as increasing row_log_interval has no effect in this case.

Correct me if I'm wrong, but in that case the training metrics are NEVER logged. After this patch, they are logged once per epoch. If they won't be logged even once per epoch, what can be the user's intention to log metrics in the training loop in the first place?

**Comment by awaelchli** [comment]
@ozen yes, they are never logged, but an alternative could be to still log every row_log_interval steps, just based on global step as suggested here #3466. in that case, when a bottleneck occurs, the user can still log every few epochs by adjusting row_log_interval.

**Comment by ozen** [comment]
Indeed, if #3466 was added it would solve this issue.

**Comment by Borda** [comment]
@ozen how about this one, is it still WIP or ready to review? 🐰

**Comment by ozen** [comment]
> 
> 
> @ozen how about this one, is it still WIP or ready to review? 🐰

The code is complete, but I didn't have time to write the test @SkafteNicki has asked.

**Comment by awaelchli** [comment]
@ozen we recently changed logging intervals to be based on global step. 
As far as I can see, the logging now always happens on the epoch (last batch) and on the interval defined by row_log_interval. 
I think this is the desired behaviour. Does this PR add anything else?

**Comment by ozen** [comment]
@awaelchli No, it doesn't add anything else.

**Pull request closed without merging**